### PR TITLE
Refactor `WriterT`-style logging

### PR DESCRIPTION
**Pull request by parsonsmatt**
Based on #1541 

This PR refactors the `WriterT` logging to use the `CPS` version. The `MonadWriter` interface is scrapped in favor of a custom class for reporting errors. `ErrMsg` is now `Builder` instead of `String`.

Overall this is a pretty decent improvement in performance. Testing this on the `persistent` repository and `persistent-test` in particular (since that package has a ton of exports and not a lot of docs). You have to remember to do the step twice, because it'll build `persistent` too on the first time due to configuration changing.

# plain 

With `cabal haddock persistent-test --haddock-options "+RTS -s -RTS"`:

```
  41,908,794,344 bytes allocated in the heap
   6,099,171,264 bytes copied during GC
     305,718,464 bytes maximum residency (20 sample(s))
       7,396,160 bytes maximum slop
             744 MiB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0      9993 colls,     0 par    3.845s   3.851s     0.0004s    0.0039s
  Gen  1        20 colls,     0 par    1.914s   1.915s     0.0958s    0.2327s

  TASKS: 5 (1 bound, 4 peak workers (4 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.001s  (  0.000s elapsed)
  MUT     time   16.285s  ( 19.616s elapsed)
  GC      time    5.759s  (  5.766s elapsed)
  EXIT    time    0.002s  (  0.008s elapsed)
  Total   time   22.046s  ( 25.390s elapsed)

  Alloc rate    2,573,483,835 bytes per MUT second

  Productivity  73.9% of total user, 77.3% of total elapsed

Documentation created:
/home/matt/Projects/persistent/dist-newstyle/build/x86_64-linux/ghc-9.4.2/persistent-test-2.13.1.3/doc/html/persistent-test/index.html
cabal haddock persistent-test --haddock-options "+RTS -s -RTS"  24.80s user 1.27s system 100% cpu 25.974 total
```

# This branch

`cabal haddock persistent-test --with-haddock=/home/matt/.cabal/bin/haddock --haddock-options "+RTS -s -RTS"`

```
  25,263,032,248 bytes allocated in the heap
   4,645,938,312 bytes copied during GC
     309,366,440 bytes maximum residency (18 sample(s))
       7,397,720 bytes maximum slop
             716 MiB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0      6017 colls,     0 par    2.607s   2.611s     0.0004s    0.0039s
  Gen  1        18 colls,     0 par    1.515s   1.515s     0.0842s    0.2277s

  TASKS: 5 (1 bound, 4 peak workers (4 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.000s  (  0.000s elapsed)
  MUT     time   11.759s  ( 13.185s elapsed)
  GC      time    4.122s  (  4.126s elapsed)
  EXIT    time    0.001s  (  0.009s elapsed)
  Total   time   15.883s  ( 17.320s elapsed)

  Alloc rate    2,148,472,781 bytes per MUT second

  Productivity  74.0% of total user, 76.1% of total elapsed

Documentation created:
/home/matt/Projects/persistent/dist-newstyle/build/x86_64-linux/ghc-9.4.2/persistent-test-2.13.1.3/doc/html/persistent-test/index.html
cabal haddock --with-haddock=/home/matt/.cabal/bin/haddock persistent-test    16.77s user 1.17s system 100% cpu 17.933 total
```

This change alone results in a pretty nice performance improvement. 
8 seconds faster to build docs for `persistent-test`, down to 18 seconds from 26 seconds - 30% improvement.
Total memory use is ~4% better (not huge, but not nothing).
Allocations are *way* down (42GB to 25GB).

### CODE REVIEW

**Comment by alexbiehl** [comment]
Nice improvement! Though I think the new dependencies will be a problem as they are not boot libraries and as such GHC can’t bootstrap them. 

Is it possible to use the bytestring builder and inline the dlist stuff?

**Comment by parsonsmatt** [comment]
Ah, that's unfortunate! Is there a way to statically ensure that only boot packages are allowed?

I can use the `Builder` that's in `ByteString`, inline `DList`, and pick the later `mtl` which has that module - `cabal` was giving mysterious errors about it due to the `cabal.project`'s hackage index state field.

**Comment by parsonsmatt** [comment]
Ah, can't have a newer `mtl` just yet due to the GHC boot package requirement. CPPed in the instance definition.

File: haddock-api/src/Haddock/Types.hs [hunk 1]
```diff
@@ -640,8 +653,49 @@ data SinceQual
...
+    deriving newtype (Semigroup, Monoid)
+
+runErrMsgM :: ErrMsgM a -> (a, ErrorMessages)
+runErrMsgM = runWriter . unErrMsgM
+
+singleMessage :: Builder -> ErrorMessages
+singleMessage = ErrorMessages . pure
+
+errorMessagesToList :: ErrorMessages -> [Builder]
+errorMessagesToList = DList.toList . unErrorMessages
```
Comment (alexbiehl) [inline haddock-api/src/Haddock/Types.hs:698]: You did a great job abstracting the error message business here. I think it'd be neat and tidy still if you'd inline the DList ala

```
newtype ErrorMessages = ErrorMessages { unErrorMessages :: [Builder] -> [Builder] }
    deriving newtype (Semigroup, Monoid)

singleMessage message = ErrorMessages . (message :)

errorMessagesToList (ErrorMessages messages) = messages []
```

reducing the amount of new code tremendously. WDYT?

Comment (parsonsmatt) [reply haddock-api/src/Haddock/Types.hs]: Good call! Ends up being just as fast, too. :)

**Comment by parsonsmatt** [comment]
Noticed that this PR *also* did a `nub => ordNub` translation. Tested it out and that was responsible for an inordinate amount of the benefit. I took the liberty of removing the last few `nub` calls. No big difference on the `persistent` test.

File: haddock-api/src/Haddock.hs [hunk 2]
```diff
@@ -461,7 +461,7 @@ render log' dflags unit_state flags sinceQual qual ifaces packages extSrcMap = d
             ppJsonIndex odir sourceUrls' opt_wiki_urls
                         unicode Nothing qual
                         ifaces
-                        ( nub
+                        ( ordNub
                         . map fst
                         . filter ((== Visible) . piVisibility . snd)
                         $ packages)
```
Comment (parsonsmatt) [inline haddock-api/src/Haddock.hs:467]: This `nub` call in particular is troubling - `packages` is probably a very large list in some cases.

File: haddock-api/src/Haddock/Types.hs [hunk 3]
```diff
@@ -640,9 +651,49 @@ data SinceQual
...
+type ErrMsg = Builder
 
+errMsgFromString :: String -> ErrMsg
+errMsgFromString = fromString
+
+errMsgToString :: ErrMsg -> String
+errMsgToString = Text.unpack . Text.decodeUtf8 . BSL.toStrict . toLazyByteString
+
+errMsgUnlines :: [ErrMsg] -> ErrMsg
+errMsgUnlines = List.foldl' (\acc x -> acc <> (x <> charUtf8 '\n')) mempty
```
Comment (alexbiehl) [inline haddock-api/src/Haddock/Types.hs:663]: Does this also do a reverse? If that’s not intended I’d suggest using an intersperse.

Comment (parsonsmatt) [reply haddock-api/src/Haddock/Types.hs]: Good call, thanks for the suggestion!

**Review by alexbiehl — APPROVED** [review]
Nice let’s go!

**Pull request merged**